### DISCUSSION TITLE
Suggestion: Add mapping generator for classes

### DISCUSSION THREAD

**Discussion by kotylo** [community/community#7870] [Copilot Conversations] [answered]
It would be nice to have a mapping feature in CoPilot. Currently, having two classes:
From.cs:
```
public enum From {
    FromOne,
    FromThree,
    FromEnd
}
```

To.cs
```
public enum To {
    One,
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    End
}
```

and wanting to generate a method that would map From into To, won't really make it faster. For example:

Following generates just new class, which is not what we want:
![image](https://user-images.githubusercontent.com/2077668/142477702-d59b08e7-6510-4846-83b2-7d268989afab.png)

Trying to be more specific yields in the generic solution:
```
// generate a function that returns maps enum From to enum To
//public static Dictionary<From, To> GenerateMap<From, To>()
//{
//    var map = new Dictionary<From, To>();
//    foreach (var item in Enum.GetValues(typeof(From)))
...
```
(which won't work, because it tries to map it one to one)

Also helping it a bit doesn't provide correct values:
![image](https://user-images.githubusercontent.com/2077668/142479092-c3e24bb4-7f1c-429f-a8a7-5dc613f631ab.png)
(there are no A, B, Cs in enum)

Helping even further it understands what I want, but the values are still imaginary:
![image](https://user-images.githubusercontent.com/2077668/142479690-1cf5a048-777d-4698-ac2b-b7ed9df5b9bc.png)
(there is no From**Two** in enum above)

--

So far I can see Copilot may analyse current file, but it won't go deep into different files to see their actual values. Or maybe it does, but the functionality to understand them is not there. 

Would be nice to have such feature, that would help with every day mapping. Maybe there could already be some extension that does just that, but enhancing Copilot with different features seems like a good idea to me :)

**Comment by MulleDK19** [comment, marked as answer]
It can already do this, but it can only see what's in the current file, from the beginning to the caret position. So copy the enums into the file, before the method, and it'll do it. Then delete them again. This is a general shortcoming of GPT-3.

**Reply by MulleDK19** [reply]
Even suggested the extension method on its own
![image](https://user-images.githubusercontent.com/3169809/142722807-c80c08d3-5426-49e4-a13e-b079217be091.png)

**Reply by MulleDK19** [reply]
It's very good at this kind of stuff. I didn't even have to suggest this one. I wanted a reverse method, and that's what it happened to suggest all on its own.
![image](https://user-images.githubusercontent.com/3169809/142722860-7e4cab0d-3b2d-46fe-b0f2-f94e8fde58c6.png)

Obviously, the last enum value was supposed to be called BackwardsAndForwards.

**Reply by kotylo** [reply]
Nice workaround, thanks for the tip. Not that faster for small entities, if it doesn't try the includes files, but shouldn't be a problem to do that for larger ones... let's hope for a GPT-4 or whatever to be even smarter :)
